// zephyr-agent@0.0.50/dist/lib/edge-actions/ze-enable-snapshot-on-pages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zeEnableSnapshotOnPages = zeEnableSnapshotOnPages;
const get_application_configuration_1 = require("../edge-requests/get-application-configuration");
const logging_1 = require("../logging");
const http_request_1 = require("../http/http-request");
const errors_1 = require("../errors");
async function zeEnableSnapshotOnPages({ pluginOptions: { application_uid }, envs_jwt, pages_url, }) {
    (0, logging_1.ze_log)('Enabling snapshot on cloudflare pages');
    (0, logging_1.ze_log)(`Uploading envs to Zephyr, for ${application_uid}`);
    const type = 'pages';
    const json = JSON.stringify(Object.assign(Object.assign({}, envs_jwt), { pages_url }));
    const { EDGE_URL, jwt } = await (0, get_application_configuration_1.getApplicationConfiguration)({
        application_uid,
    });
    const [ok, cause] = await (0, http_request_1.makeRequest)({
        path: '/upload',
        base: EDGE_URL,
        query: { type },
    }, {
        method: 'POST',
        headers: {
            'Content-Length': Buffer.byteLength(json).toString(),
            'Content-Type': 'application/json; charset=utf-8',
            can_write_jwt: jwt,
        },
    }, json);
    if (!ok) {
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_DEPLOY_LOCAL_BUILD, {
            cause,
            message: 'Failed to upload envs to Zephyr',
        });
    }
    (0, logging_1.ze_log)('Build successfully deployed.');
}
