// zephyr-agent@0.0.50/dist/lib/logging/debug.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ze_log = exports.brightRedBgName = exports.brightGreenBgName = exports.brightYellowBgName = exports.brightBlueBgName = exports.dimmedName = void 0;
// light weight functions for decorated console.error + able to toggle different parts of whole module's logging result
const debug_1 = require("debug");
const picocolor_1 = require("./picocolor");
//TODO: this should be traced and logged into new relic
const name = ' ZEPHYR ';
exports.dimmedName = (0, picocolor_1.dim)(name);
exports.brightBlueBgName = (0, picocolor_1.bold)((0, picocolor_1.bgCyanBright)((0, picocolor_1.black)(name)));
exports.brightYellowBgName = (0, picocolor_1.bold)((0, picocolor_1.bgYellowBright)((0, picocolor_1.black)(name)));
exports.brightGreenBgName = (0, picocolor_1.bold)((0, picocolor_1.bgGreenBright)((0, picocolor_1.black)(name)));
exports.brightRedBgName = (0, picocolor_1.bold)((0, picocolor_1.bgRedBright)((0, picocolor_1.black)(name)));
exports.ze_log = (0, debug_1.debug)('zephyr:log');
// If debug mode is not enabled just print whatever console output is
// If debug mode is enabled print the error from our end
